osu.Server.Spectator/Hubs/EntityStore.cs

```diff
@@ -0,0 +1,165 @@
+
+                    entityMapping[id] = item = new TrackedEntity(id, this);
+                }
+            }
+
+            await item.ObtainLockAsync();
+
+            return new ItemUsage<T>(item);
+        }
+
+        public async Task Destroy(long id)
+        {
+            TrackedEntity? item;
+
+            lock (entityMapping)
+            {
+                if (!entityMapping.TryGetValue(id, out item))
+                    // was not tracking.
+                    return;
+            }
+
+            await item.ObtainLockAsync();
+
+            // handles removal and disposal of the semaphore.
+            item.Destroy();
+        }
+
+        /// <summary>
+        /// Get all tracked entities in an unsafe manner. Only read operations should be performed on retrieved entities.
+        /// </summary>
+        public KeyValuePair<long, T>[] GetAllEntities()
+        {
+            lock (entityMapping)
+            {
+                return entityMapping
+                       .Where(kvp => kvp.Value.GetItemUnsafe() != null)
+                       .Select(entity => new KeyValuePair<long, T>(entity.Key, entity.Value.GetItemUnsafe().AsNonNull()))
+                       .ToArray();
+            }
+        }
+
+        /// <summary>
+        /// Clear all tracked entities.
+        /// </summary>
+        public void Clear()
+        {
+            lock (entityMapping)
+            {
+                entityMapping.Clear();
+            }
+        }
+
+        private void remove(long id)
+        {
+            lock (entityMapping)
+                entityMapping.Remove(id);
+        }
+
+        public class TrackedEntity
+        {
+            private readonly SemaphoreSlim semaphore = new SemaphoreSlim(1);
+
+            private T? item;
+
+            private readonly long id;
+            private readonly EntityStore<T> store;
+
+            private bool isDestroyed;
+
+            private bool isLocked => semaphore.CurrentCount == 0;
+
+            public TrackedEntity(long id, EntityStore<T> store)
+            {
+                this.id = id;
+                this.store = store;
+            }
+
+            public T? GetItemUnsafe() => item;
+
+            public T? Item
+            {
+                get
+                {
+                    checkValidForUse();
+                    return item;
+                }
+                set
+                {
+                    checkValidForUse();
+                    item = value;
+                }
+            }
+
+            /// <summary>
+            /// Mark this item as no longer used. Will remove any tracking overhead.
+            /// </summary>
+            public void Destroy()
+            {
+                // we should already have a lock when calling destroy.
+                checkValidForUse();
+
+                isDestroyed = true;
+
+                store.remove(id);
+                semaphore.Dispose();
```

Comment:
I figured out why my test was failing here. It appears that it is in fact not safe to dispose a semaphore that other readers are waiting on. And by that I don't mean "it'll timeout", I mean "it will block forevermore". The [docs](https://docs.microsoft.com/en-us/dotnet/api/system.threading.semaphoreslim.dispose?view=net-5.0#System_Threading_SemaphoreSlim_Dispose) slightly hint at this:

> Call `Dispose` when you are finished using the `SemaphoreSlim`. The `Dispose` method leaves the `SemaphoreSlim` in an unusable state.

but I think they should be shouting that from the rooftops rather than this meek mention... If you compare the implementation of `Dispose()` and `Release()` it does look like the former just bails instead of cleaning up properly.

In light of that I think it's proper to release the semaphore before doing this here. It instantly fixes my test, for what it's worth.